src/client.ts

```diff
@@ -0,0 +1,217 @@
+const MAX_SEEN_TRACKING = 2048;
+
+export type StartStreamResponse = {
+  token: string;
+  channel_id: string;
+  algorithm: string;
+  expires_in: number;
+};
+
+export type BaseEnvelope = {
+  event_type: 'stream' | 'control';
+  message_id: string;
+};
+
+// To match the rust enums
+export type StreamPhase = 'PHASE_START' | 'PHASE_DELTA' | 'PHASE_END' | 'PHASE_ERROR';
+
+export type StreamEnvelope<T> = BaseEnvelope & {
+  event_type: 'stream';
+  sequence: number;
+  phase: StreamPhase;
+  payload?: T;
+};
+
+export type ControlEnvelope = BaseEnvelope & {
+  event_type: 'control';
+  control_type: 'server_draining';
+};
+
+export type ConduitClientConfig<T> = {
+  orgId: number;
+  streamUrl: string;
+  baseConduitUrl: string;
+  startStreamData?: Record<string, unknown>;
+  onMessage?: (message: T) => void;
+  onOpen?: () => void;
+  onClose?: () => void;
+  onError?: (err: Error) => void;
+};
+
+export class ConduitClient<T> {
+  private config: ConduitClientConfig<T>;
+  private eventSource: EventSource | null = null;
+  private connecting = false;
+
+  private lastEventId: string | undefined;
+  private lastSeq: number | undefined;
+  private seenIds = new Set<string>();
+
+  private tokenExpiresAt: number | undefined;
+  private currentChannelId: string | undefined;
+
+  constructor(config: ConduitClientConfig<T>) {
+    this.config = config;
+    this.handleStream = this.handleStream.bind(this);
+    this.handleControl = this.handleControl.bind(this);
+  }
+
+  private async startStream(): Promise<StartStreamResponse> {
+    const response = await fetch(this.config.streamUrl, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify({
+        org_id: this.config.orgId,
+        ...this.config.startStreamData,
+      }),
+    });
+
+    if (!response.ok) {
+      throw new Error(`Failed to start stream ${response.statusText}`);
+    }
+
+    return response.json();
+  }
+
+  private buildUrl(token: string, channelId: string): string {
+    const queryParams = new URLSearchParams({
+      token,
+      channel_id: channelId,
+    });
+    if (this.lastEventId) queryParams.set('last_event_id', this.lastEventId);
+    return `${this.config.baseConduitUrl.replace(/\/$/, '')}/events/${this.config.orgId}?${queryParams.toString()}`;
+  }
+
+  private attach(url: string) {
+    if (this.tokenExpiresAt && Date.now() >= this.tokenExpiresAt) {
+      return;
+    }
```

Comment:
If the token is expired then we just return from attach? Should this be an error?